### PR TITLE
[chip_sw, sysrst_ctrl] Test ulp wakeup of sysrst_ctrl

### PR DESCRIPTION
**Pull request by ballifatih**
This is a simple test case for ulp wakeup. The test passes on dvsim, and I could use some suggestions to improve it.

### CODE REVIEW

**Review by msfschaffner — COMMENTED** [review]
Thanks @ballifatih for adding this!

I have added @weicaiyang and @sriyerg to help take a look, since they may have some additional feedback regarding the DV portion.

We should probably extend this testpoint and also put the chip into sleep mode before forcing the lid-open pin. After all, this is testing a wakeup feature and we would like to make sure that the communication with the power manager is working correctly. @sriyerg, @weicaiyang WDYT?

This should not be too hard to implement as it would mainly need some additional power manager initialization steps, similar to what is done in the the `chip_sw_adc_ctrl_sleep_debug_cable_wakeup` and `chip_sw_sysrst_ctrl_reset` tests.

File: sw/device/tests/sim_dv/sysrst_ctrl_ulp_z3_wakeup_test.c
Comment (msfschaffner) [inline sw/device/tests/sim_dv/sysrst_ctrl_ulp_z3_wakeup_test.c:19]: does this variable still exist?

File: hw/top_earlgrey/dv/env/seq_lib/chip_sw_sysrst_ctrl_ulp_z3_wakeup_vseq.sv
Comment (weicaiyang) [inline hw/top_earlgrey/dv/env/seq_lib/chip_sw_sysrst_ctrl_ulp_z3_wakeup_vseq.sv:32]: you probably follow the other sysrst_ctrl test to force the IO.
We already re-constructed IO connection in chip_if, and now we have cfg.chip_vif.sysrst_ctrl_if which allows to drive sysrst_ctrl without forcing. You can refer to this seq.
https://cs.opensource.google/opentitan/opentitan/+/master:hw/top_earlgrey/dv/env/seq_lib/chip_sw_sysrst_ctrl_outputs_vseq.sv

**Comment by sriyerg** [comment]
The sysrst_ctrl IOs are incorrectly connected to random pins in the chip. Plus, the existing sysrst ctrl tests employed forces for controlling the pins. We have now refactored the logic of how we connect external interfaces to the chip IOs, which eliminates the need to use forces. Please hold off if possible, on merging this PR, until the rest of the sysrst ctrl tests are updated to use the new flow (which will give you the means to properly develop this test) and are connected to the right chip pins.

**Comment by sriyerg** [comment]
> Thanks @ballifatih for adding this!
> 
> I have added @weicaiyang and @sriyerg to help take a look, since they may have some additional feedback regarding the DV portion.
> 
> We should probably extend this testpoint and also put the chip into sleep mode before forcing the lid-open pin. After all, this is testing a wakeup feature and we would like to make sure that the communication with the power manager is working correctly. @sriyerg, @weicaiyang WDYT?
> 
> This should not be too hard to implement as it would mainly need some additional power manager initialization steps, similar to what is done in the the `chip_sw_adc_ctrl_sleep_debug_cable_wakeup` and `chip_sw_sysrst_ctrl_reset` tests.

Agreed - it will be good to update the testplan and the test to also verify wake up from deep sleep state.

File: sw/device/tests/sim_dv/sysrst_ctrl_ulp_z3_wakeup_test.c
Comment (ballifatih) [reply sw/device/tests/sim_dv/sysrst_ctrl_ulp_z3_wakeup_test.c]: `kTestExpected` is not used, so I removed it. Thanks for catching that.

**Comment by ballifatih** [comment]
@weicaiyang I initially tried with `sysrst_ctrl_if` but I couldn't get it to work. I can make the transition to `chip_if` when @sriyerg merges the pin configuration update. Thanks for the suggestion and clarification.

Meanwhile, I think we can continue updating this PR. I have made a small change that also puts the chip into deep sleep. It would be great if someone can have a look, because I didn't fully grasp the interaction between pwrmgr and rstmgr. @msfschaffner 

This is what I'm doing with the test after the update:

1. Set lid_open = 0
2. Check there is no wake up event
3. Enter deep sleep mode
4. Start glitching lid_open for a short duration (less than debounce timer) and then stablize on '1'
5. Wait sw to wake up
6. Check if the z3_wakeup pin is raised

File: sw/device/tests/sim_dv/sysrst_ctrl_ulp_z3_wakeup_test.c
Comment (msfschaffner) [inline sw/device/tests/sim_dv/sysrst_ctrl_ulp_z3_wakeup_test.c:123]: Since we also initialize the reset manager, we should probably check that the reset reason is due to a low-power exit `kDifRstmgrResetInfoLowPowerExit`. See e.g. here: 
https://github.com/lowRISC/opentitan/blob/9ca43a9f2dbfa676a6c5b38857eb024704bafe65/sw/device/tests/aon_timer_wdog_bite_reset_test.c#L113-L120

File: sw/device/tests/sim_dv/sysrst_ctrl_ulp_z3_wakeup_test.c
Comment (msfschaffner) [inline sw/device/tests/sim_dv/sysrst_ctrl_ulp_z3_wakeup_test.c:119]: The low-power setup here LGTM.

File: sw/device/tests/sim_dv/sysrst_ctrl_ulp_z3_wakeup_test.c
Comment (msfschaffner) [inline sw/device/tests/sim_dv/sysrst_ctrl_ulp_z3_wakeup_test.c:84]: Can you use docstring format for all these functions to enable automatic extraction via doxygen?
```
/** 
 * Waits for the kTestPhase variable to be changed by a backdoor overwrite
 * from the testbench in chip_sw_<testname>_vseq.sv. This will
 * indicate that the testbench is ready to proceed with the
 * next phase of the test.
 */
```

**Review by msfschaffner — COMMENTED** [review]
Thanks @ballifatih, I took a quick look and have a few comments.

@weicaiyang WDYT, should we iterate over the wakeup procedure multiple times in this test?

**Comment by weicaiyang** [comment]
> @weicaiyang WDYT, should we iterate over the wakeup procedure multiple times in this test?

We didn't mention that in the testplan. So either way is fine to me.

**Comment by weicaiyang** [comment]
LGTM. defer to @msfschaffner for approval.

**Comment by msfschaffner** [comment]
> LGTM. defer to @msfschaffner for approval.

OK sounds good to me (I believe we cover multiple iterations at the block-level already so this should be fine).

File: sw/device/tests/sim_dv/sysrst_ctrl_ulp_z3_wakeup_test.c
Comment (ballifatih) [reply sw/device/tests/sim_dv/sysrst_ctrl_ulp_z3_wakeup_test.c:136]: Thanks for the comments @msfschaffner. I just added this check.

**Comment by ballifatih** [comment]
The test is currently failing on dvsim, because I am not able to drive `tb.dut.XYZ` pins after rebasing on master (if I interpret my waveforms correctly). Since we will already switch to `sysrst_ctrl_if` to drive IO pads anyway, I will wait for other IO re-configurations to be completed @sriyerg . At the moment, I see some `sysrst_ctrl` tests failing on the HW dashboard.

**Comment by ballifatih** [comment]
The test is currently passing by using `sysrst_ctrl_if`, but we can wait #15811 to merge for another round of reviews.

**Review by ballifatih — COMMENTED** [review]
@msfschaffner @weicaiyang 
Unless you have new comments, I think this test should be OK to merge. Two changes from last time:

1. This test now uses sysrst_ctrl pins to drive IO. 
2. It uses the updated sw-backdoor/polling to sync dv and sw, as merged in #15811.

The only question is whether it is OK to drive low the input pins instead of completely disconnecting them in `post_start`. What do you think @sri?

File: hw/top_earlgrey/dv/env/seq_lib/chip_sw_sysrst_ctrl_ulp_z3_wakeup_vseq.sv
Comment (ballifatih) [inline hw/top_earlgrey/dv/env/seq_lib/chip_sw_sysrst_ctrl_ulp_z3_wakeup_vseq.sv:38]: Is it OK to drive pins low instead of disconnecting them at the end of the test? Disconnecting pins cause [`PwrbOKnown_A` assertion](https://cs.opensource.google/opentitan/opentitan/+/master:hw/ip/sysrst_ctrl/rtl/sysrst_ctrl.sv;l=390?q=PwrbOKnown_A&ss=opentitan) to fail. WDYT @sriyerg ?

File: sw/device/tests/sim_dv/sysrst_ctrl_ulp_z3_wakeup_test.c
Comment (moidx) [inline sw/device/tests/sim_dv/sysrst_ctrl_ulp_z3_wakeup_test.c:123]: nit: To follow the style guide, remove spaces, i.e.: `/*pwrmgr_domain_config=*/ 0`

**Review by moidx — APPROVED** [review]

**Review by weicaiyang — APPROVED** [review]
LGTM

File: sw/device/tests/sim_dv/sysrst_ctrl_ulp_z3_wakeup_test.c
Comment (msfschaffner) [inline sw/device/tests/sim_dv/sysrst_ctrl_ulp_z3_wakeup_test.c:169]: Hm I don't think this is this the right place to check for the wakeup reason.
If we are going to deep sleep, the core will be shut off and reset, meaning that we will never execute this...

It's probably better to do the following:
- modify `check_wakeup_reason()` to `reset_is_low_power_exit()` and return a bool
- add `CHECK(!reset_is_low_power_exit());` before `CHECK(!reset_is_low_power_exit());`
- add `CHECK(reset_is_low_power_exit());` before `CHECK(reset_is_low_power_exit());`

**Review by msfschaffner — COMMENTED** [review]
Thanks for updating the SV side, @ballifatih!
One more comment on checking the reset reason, otherwise this LGTM.

File: sw/device/tests/sim_dv/sysrst_ctrl_ulp_z3_wakeup_test.c
Comment (msfschaffner) [inline sw/device/tests/sim_dv/sysrst_ctrl_ulp_z3_wakeup_test.c:172]: This looks good, thanks for adding it!

**Review by msfschaffner — APPROVED** [review]

**Pull request merged**
